compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeSceneMediator.desktop.kt

```diff
@@ -342,6 +344,15 @@ internal class ComposeSceneMediator(
         // to react only on changes with [interopLayer].
         container.addContainerListener(containerListener)
 
+        // AwtDragAndDropManager support
+        container.transferHandler = ComposeTransferHandler(container)
+        container.dropTarget = ComposeDropTarget(
+            rootContainer = container,
+            dropTargetListener = {
+                (scene.activeDragAndDropManager() as? AwtDragAndDropManager)?.dropTargetListener
```

Comment:
I polished and pushed the patch, but I really think we should discuss the necessity of decoupling our desktop code from the AWT implementation.

It significantly complicates our code and data flow and sometimes even negatively affects the public API we present. The benefits are mostly illusory in my opinion (but maybe I'm wrong). Anyone needing to provide an alternative implementation will be better served by a simpler codebase, even if they need to recompile the project.

Reply:
> necessity of decoupling our desktop code from the AWT implementation.

There are 2 cases of `desktopMain`:

1. Using AWT in `foundation` and higher level modules.

In this case we are trying to eliminate `desktopMain` sourceSet at all, an move all code to `skikoMain` (or better to `commonMain`). It is beneficial for usual users of Compose Multiplatform when they write a cross-paltform application. 

Let's remember a real situation with `AlertDialog` that had own implementations for each platform. `AlertDialog` is usually very opinionated thing, and if we need to customize it, we have to copy paste code. But because it had completely different implementations on all platforms, we had to copy-paste it on each platform.

Now we commonized it, extracting a public API `Dialog` in `ui`. We solved 2 issues - users only need to copy paste `commonMain`, and they have a public platform abstraction for `Dialog`, they can build on top of it very customised dialogs. Just extracting this abstraction into `expect` won't be enough, as it will be in `material` only, we need it in `ui` (or a hypothetical future `ui-platform`), so other design system can access it.

`androidMain` currently don't follow this principle as it uses Android SDK directly, and has this copy-paste issue. To follow this principle, we need to move all the code to `commonMain` and design a lower-level platform facade in `ui`. This is the price of having platform abstraction over lower level.

2. Using AWT in `ui` module. 

We have 2 options here:
- create an `expect` and use AWT directly
- create an `interface` and inject it into `ComposeScene`

In my opinion these are just 2 different ways to provide an implementation, and the second one is not more difficult than the first one.

But the second one has advantages:
- allowing constructing a custom platform bindings. I remember real user cases where it was useful:
  - Compose rendering on a server
  - Virtual testing
  - Replacing AWT by JWM, which behave better in window management things.
  - Minecraft integration 🙃

- it separates Compose and a platform, which are on different layers of abstractions. Mixing different layers usually leads to over-complicated code if people aren't careful. Having separation by design helps to avoid this, and helps to think about new features in a certain way (is it a Compose feature, or is it a platform feature?)

See also the recent presentation, it partially covers why we doing it this way.